src/utransport.rs

```diff
@@ -12,13 +12,122 @@
+/// `UListener` is the uP-L1 interface that provides a means to create listeners which are registered to [`UTransport`]
 ///
-/// Implementations of `UTransport` contain the details for connecting to the underlying transport technology and
-/// sending `UMessage` using the configured technology. For more information, please refer to
+/// Implementations of `UListener` contain the details for what should occur when a message is received
+///
+/// For more information, please refer to
+/// [uProtocol Specification](https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/up-l1/README.adoc).
+///
+/// # Examples
+///
+/// ## Simple example
+///
+/// ```
+/// use up_rust::{UListener, UMessage, UStatus};
+///
+/// use async_trait::async_trait;
+/// use std::sync::{Arc, Mutex};
+///
+/// #[derive(Clone)]
+/// struct FooListener {
+///     inner_foo: Arc<Mutex<String>>
+/// }
+///
+/// #[async_trait]
+/// impl UListener for FooListener {
+///     async fn on_receive(&self, msg: UMessage) {
+///         let mut inner_foo = self.inner_foo.lock().unwrap();
+///         if let Some(payload) = msg.payload.as_ref() {
+///             if let Some(length) = payload.length.as_ref() {
+///                 *inner_foo = format!("latest message length: {length}");
+///             }
+///         }
+///     }
+///
+///     async fn on_error(&self, err: UStatus) {
+///         println!("uh oh, we got an error: {err:?}");
+///     }
+/// }
+/// ```
+///
+/// ## Long-running function needed when message received
+///
+/// ```
+/// use up_rust::{UListener, UMessage, UStatus};
+///
+/// use async_trait::async_trait;
+/// use async_std::task;
+/// use std::sync::{Arc, Mutex};
+///
+/// #[derive(Clone)]
+/// struct LongTaskListener;
+///
+/// async fn send_to_jupiter(message_for_jupiter: UMessage) {
+///     // send a message to Jupiter
+///     println!("Fly me to the moon... {message_for_jupiter}");
+/// }
+///
+/// #[async_trait]
+/// impl UListener for LongTaskListener {
+///     async fn on_receive(&self, msg: UMessage) {
+///         task::spawn(send_to_jupiter(msg));
+///     }
+///
+///     async fn on_error(&self, err: UStatus) {
+///         println!("unable to send to jupiter :( {err:?}");
+///     }
+/// }
+/// ```
+#[async_trait]
+pub trait UListener: 'static + Send + Sync {
+    /// Performs some action on receipt of a message
+    ///
+    /// # Parameters
+    ///
+    /// * `msg` - The message
+    ///
+    /// # Note for `UListener` implementers
+    ///
+    /// `on_receive()` is expected to return almost immediately. If it does not, it could potentially
+    /// block further message receipt. For long-running operations consider passing off received
+    /// data to a different async function to handle it and returning.
+    ///
+    /// # Note for `UTransport` implementers
+    ///
+    /// Because `on_receive()` is async you may choose to either `.await` it or spawn it onto its
```

Comment:
you would still need to await it, otherwise no work will be done, if I am not mistaken ...

Reply:
> you would still need to await it, otherwise no work will be done, if I am not mistaken ...

You're 100% right... I didn't word this very well. Let me go clean this up.